### PR TITLE
CI: Run tests in a single CI job (Ubuntu + Python 3.9) for draft PRs

### PR DESCRIPTION
**Pull request by seisman**
**Description of proposed changes**

This PR is inspired by another PR #869 from @willschlitzer.

This PR improves [the "Tests" workflow](https://github.com/GenericMappingTools/pygmt/blob/master/.github/workflows/ci_tests.yaml) to save CI resources. 

Expected behavior (see the comments below for how it actually works):

- For draft PRs: only run tests on Ubuntu + Python 3.9 (~10 minutes)
- When draft PRs are ready for review, run all CI jobs (9 jobs, 10-30 minutes)
- If converting PRs back to draft mode, pushes still only trigger one job (~10 minutes)
- After merged into master, all 9 jobs will run.

References:

1. https://github.com/asyncapi/parser-js/pull/204
2. https://github.community/t/how-to-conditionally-include-exclude-items-in-matrix-eg-based-on-branch/16853/6

Closes #869.

### CODE REVIEW

**Comment by seisman** [comment]
Now, this PR is in draft mode. As expected, commit af67396483cff0f84469c0e609bdc71898bcb384 only triggers one job (Ubuntu + Python 3.9). Check https://github.com/GenericMappingTools/pygmt/actions/runs/570286139 for the job details. Below is the screenshot of the CI runs. Only the "ubuntu-latest - Python 3.9" job is triggered. Other jobs are waiting for status because they're not triggered.

image

**Comment by seisman** [comment]
This PR is still in draft mode. New commits like 7c6309d2111bb2779c3b0d948ba8deca8de5ee27 still only triggers one job (ubuntu-latest - Python 3.9). Check https://github.com/GenericMappingTools/pygmt/actions/runs/570313438 for job details. Here is the screenshot of the CI runs.


TODO: add the screenshot here.

**Comment by seisman** [comment]
I just pushed the "Ready for Review" button, which triggers all the CI jobs. Check https://github.com/GenericMappingTools/pygmt/actions/runs/570319849 for job details.

Below is the screenshot for job status. Some notes:

- All 9 jobs are triggered now, and the old "ubuntu-latest - Python 3.9" job is not canceled. So 10 jobs have been triggered by the commit and the "ready_for_review" event.
- The "Style Checks" workflow is not re-triggered by "ready for review" event.
- Three jobs in the "GMT Dev Tests" workflow are also triggered by the "ready_for_review" event.

image

**Comment by seisman** [comment]
The PR is not in draft mode, so new commits like 4dc7e87d1cefe9b0e8d9dfdc25850c91101e6a53 trigger new CI runs (9 jobs). Check https://github.com/GenericMappingTools/pygmt/actions/runs/570349410 for job details.

Some notes:

- The new commit triggers new CI runs (9 jobs) in the "Tests" workflow
- This new commit does not trigger any runs in the "GMT Dev Tests" workflow, because the workflow is only triggered by the `ready_for_review` event.

image

**Comment by seisman** [comment]
PR authors and project maintainers can convert a PR into draft mode. Note that "convert to draft" is not an "event", so it doesn't trigger new CI runs.

**Comment by seisman** [comment]
Now the PR is in draft mode. New commits like 06d00734386c6809b91704e22c0f1a01db97f95e only triggers one job (ubuntu-latest + Python 3.9). Check https://github.com/GenericMappingTools/pygmt/actions/runs/570386483 for job details.

image

**Comment by seisman** [comment]
Now I can make more changes in draft mode. Each commit only triggers one job, so saving us 8/9 CI resources, especially for macOS and Windows.

Like above, if I mark the PR as ready for review again, 9 jobs will be triggered.

So what will happen if the changes are merged into master branch? I can't show you in this repository. Instead, I made similar changes to my own fork (https://github.com/seisman/pygmt), and merge [the changes](af92685bda13f0a20c945ef0f4367e96d3a2ae12) into my master branch. Everything works well. See the screenshot below and https://github.com/seisman/pygmt/actions/runs/570241183 for job details.

The "Publish to PyPI" workflow fails as reported in #901.

image

**Comment by seisman** [comment]
Now this PR is ready for review.

File: MAINTENANCE.md
Comment (weiji14) [inline MAINTENANCE.md:62]:
```suggestion
In draft Pull Requests, only one job (Ubuntu + Python latest)
is triggered to save on Continuous Integration resources.
```
Use full text.

File: .github/workflows/ci_tests.yaml
Comment (weiji14) [inline .github/workflows/ci_tests.yaml:33]:
```suggestion
        # Is it a draft Pull Request (true or false)?
```

File: .github/workflows/ci_tests.yaml
Comment (weiji14) [inline .github/workflows/ci_tests.yaml:10]: Could you explain these 'types' a bit more? I understand opened, [reopened](https://docs.github.com/en/developers/webhooks-and-events/issue-event-types#reopened) and [ready_for_review](https://docs.github.com/en/developers/webhooks-and-events/issue-event-types#ready_for_review), but what does synchronize do? I see 'synchronize is listed on https://docs.github.com/en/developers/webhooks-and-events/github-event-types#pullrequestevent but the docs don't say what that status means.

**Review by weiji14 — COMMENTED** [review]
This is awesome, now I won't feel so bad burning through so many CPU cycles! Just a few comments.

File: .github/workflows/ci_tests.yaml
Comment (seisman) [reply .github/workflows/ci_tests.yaml:10]: `synchronize` is better explained in this issue https://github.com/github/docs/issues/2257.

FYI, the [documentation](https://docs.github.com/en/actions/reference/events-that-trigger-workflows#pull_request) says, 

> By default, a workflow only runs when a pull_request's activity type is opened, synchronize, or reopened.

Here we need `ready_for_review` to trigger the full tests when pushing the button.

Comment (weiji14) [reply .github/workflows/ci_tests.yaml:10]: Ah ok, sounds good then.

File: MAINTENANCE.md
Comment (weiji14) [inline MAINTENANCE.md:63]:
```suggestion
Alternatively, add [skip ci] to your commit message so that Github Actions won't run.
See https://github.blog/changelog/2021-02-08-github-actions-skip-pull-request-and-push-workflows-with-skip-ci/.

```

**Review by weiji14 — APPROVED** [review]
Up to you if you want to apply the changes below to close #908, otherwise good to merge.

File: MAINTENANCE.md
Comment (seisman) [reply MAINTENANCE.md:63]: Prefer to leave it to another PR, as more tests need to be done before documenting it.

Comment (weiji14) [reply MAINTENANCE.md:63]: Ok, feel free to copy this text to that next PR.

**Pull request merged**
